Update address JS utils & remove filter

ralucaStan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

- - This PR converts the `addressFieldsForShippingRates` setting introduced in #57374 to a JS constant kept in the address utils file.

-- - This PR also removes the `woocommerce_address_fields_for_shipping_rates` filter introduced in #57374. The filter is not released, so no deprecation is needed. This was an optimistically added filter. We're not accounting for it in the shipping calculator (which is pending removal https://github.com/woocommerce/woocommerce/issues/57995) and the WC shipping logic doesn't account for additional shipping fields. It was only intended for front-end use in the checkout block to trigger loading states. If a need arises we'd need to create a holistic approach at the core level

- This PR updates the address utils to account for `addressFieldsForShippingRates` setting introduced in #57374 instead of the hardcoded values for `[ 'state', 'country', 'postcode', 'city' ]`.

- The `isAddressComplete` function was renamed and refactored to use the shipping fields setting; we're not using it for checking random fields, it was used only for the fields from addressFieldsForShippingRates.




<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR does not account for the hardcoded variables on the PHP side, this would be handled in another PR.



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### a. No shipping rates defined for a complete address
1. Define shipping rates only for one country
2. Checkout with a product, and select another country
3. Notice that checkout shows a message that there are no shipping options
4. Notice the order summary show a message about missing delivery
5. Fill in the shipping address, select shipping option and checkout

#### b. Guest user with no address
1. Set in WooCommerce > General >  Default customer location > No location by default
2. Go to a guest checkout and notice the shipping area and the Order summary message prompting to enter an address
<img width="674" alt="Screenshot 2025-05-15 at 16 03 43" src="https://github.com/user-attachments/assets/e3ce3c8d-e016-4cf6-aa6f-93527c6724ed" />
3. Fill in the shipping address, select shipping option and checkout

### Shipping calculator
Testing for this is covered by this file https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/tests/e2e/tests/cart/cart-checkout-block-shipping.block_theme.spec.ts
<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ralucaStan committed:
Update address JS utils to use the addressFieldsForShippingRates setting
Refactored and renamed the isAddressComplete function to use the shipping fields setting; we're not using it for checking random fields, it was used only for the fields from addressFieldsForShippingRates
Removed some JS utils we weren't using

ralucaStan committed:
Fix typo

ralucaStan committed:
add changelog

ralucaStan committed:
Add the addressFieldsForShippingRates setting to cart

ralucaStan committed:
Remove the woocommerce_address_fields_for_shipping_rates filter

This was an optimistically added filter. We're not accounting for it in the shipping calculator and the shipping logic isn't account for additional shipping fields

ralucaStan committed:
Merge branch 'trunk' into update/is-address-complete

ralucaStan committed:
Merge branch 'trunk' into update/is-address-complete

opr commented on plugins/woocommerce/client/blocks/assets/js/types/type-defs/cart-response.ts:
```diff
@@ -163,6 +163,11 @@ export interface CartResponseExtensionItem {
+	extends Omit< ResponseBaseAddress, 'address_1' | 'address_2' > {
+	[ key: string ]: string;
+}
+
```
I don't think this type is correct because it's saying that `ShippingAddressFieldsForShippingRates` is an object of strings, keyed by strings, whereas it is an array of strings. We should probably use the keys of `ResponseBaseAddress`without `address_1` and `address_2`.

e.g. 
```suggestion
export type ShippingAddressFieldsForShippingRates = keyof Omit<
	ResponseBaseAddress,
	'address_1' | 'address_2'
>;
```

ralucaStan replied:
Nice catch! I've based it on BaseAddress, as I've learned that the response types are for data coming from Store API. See https://github.com/woocommerce/woocommerce/pull/58030/commits/f8200fec0b381e135c50465ea3a4b78a38125975

opr commented on plugins/woocommerce/src/Blocks/BlockTypes/Cart.php:
```diff
@@ -252,6 +252,8 @@ protected function enqueue_data( array $attributes = [] ) {
 		$this->asset_data_registry->add( 'collectableMethodIds', $local_pickup_method_ids );
 		$this->asset_data_registry->add( 'shippingMethodsExist', CartCheckoutUtils::shipping_methods_exist() > 0 );
 
+		$this->asset_data_registry->add( 'addressFieldsForShippingRates', [ 'state', 'country', 'postcode', 'city' ] );
```
What is the purpose of adding this to the asset_data_registry? If it's a static value, and only consumed on the front end, should it just be kept as a static value in `address.ts` ? We can remove the PHP tests checking for this value too if we do that.

ralucaStan replied:
You're right. I've moved it to the address utilities. Thank you for catching that.

opr reviewed (changes requested):
Thanks for working on this, I tested and it's working fine. I agree with removing the filters for the address fields, but I think it's safe to stop defining this in PHP if we're not filtering it. Do you agree?

If so, I can re-test it once the `addressFieldsForShippingRates` is defined in the front-end.

ralucaStan committed:
Merge branch 'trunk' into update/is-address-complete

ralucaStan committed:
Remove addressFieldsForShippingRates from settings
This is a JS only constant, so there is no need to store it in the settings via the assets registry

ralucaStan committed:
Fix type and lint error

ResponseBaseAddress relates to Store API types so I'm changing the type base.

ralucaStan committed:
Merge branch 'trunk' into update/is-address-complete

ralucaStan committed:
Merge branch 'trunk' into update/is-address-complete

# Conflicts:
#	plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/totals/shipping/shipping-address.tsx

opr reviewed (approved):
Thanks Raluca, looks good now!
